### DISCUSSION TITLE
Is it possible to protect a single page with htaccess/htpasswd?

### DISCUSSION THREAD

**Discussion by glasen** [datenstrom/community#132] [Ask a question] [answered]
Hi there,

is it possible to protect a single page with htaccess/htpasswd?

The use case is following:

Me and my fellow students have a lecture named "scripting languages for bioinformatics" and our professor wants us to upload our python files to a subpage of our homepage (The homepage i successfully got running a hour ago ;-) ). The subpage should be protected by a htpasswd-file. There is no other option because the prof gave us a copy of his hashed login-data so he can log into the subpage to look for the files or other things.

**Comment by wunderfeyd** [comment]
Hi,

you can protect whole directories in your (for example) `media/downloads/` folder, with a `.htaccess` and `.htpasswd` file combination. Note that you can't protect files inside the content folder with this.

`media/downloads/aufgaben/.htaccess`:

```
AuthType Basic
AuthName "Aufgaben"
AuthUserFile /var/www/media/downloads/aufgaben/.htpasswd
require valid-user
```

`media/downloads/aufgaben/.htpasswd`:

```
test:teH0wLIpW0gyQ
```

Please note you have to add users and adjust your folder names. Then you are able to link from within a single page in Markdown syntax to these files. Go to the directory in your browser and click `Copy location...` and use this for your links like: `[Download](http://example.com/media/downloads/aufgaben/yellow.pdf)`.

---

As an alternative if you want to protect the `content` directory, you can add another yellow user with another home location and use the [draft plugin](https://github.com/datenstrom/yellow-extensions/tree/master/plugins/draft) to protect these pages. In case of questions about this topic don't hesitate to ask.

**Comment by markseuffert** [comment]
As a third alternative, as a developer you can use the Yellow API to write your own plugins. It's would be possible to check for `Status: protected` in the settings of a page and tell the browser to provide authentication. You could even copy the hashed login-data to the settings, which means in theory you could configure everything in one place. 

Just an idea. :smile:

**Comment by glasen** [comment, marked as answer]
@markseu:
Sounds good. I already wanted to extent my PHP-skills and a "protected site"-feature sounds like a nice extension to Yellow.

BTW:

I found a solution for my problem. It is possible to protect a specific URL with htaccess:

``` html:1
AuthType Basic
AuthName "Restricted Area"
AuthUserFile /home/sgls22/public_html/.htpasswd
Require valid-user
SetEnvIf Request_URI "(vorlesungen/skriptsprachen)$" deny
Deny from env=deny
Satisfy any
```
